staging/src/k8s.io/kubectl/pkg/cmd/util/helpers.go

```diff
@@ -625,8 +634,41 @@ func DumpReaderToFile(reader io.Reader, filename string) error {
 	return nil
 }
 
-func GetServerSideApplyFlag(cmd *cobra.Command) bool {
-	return GetFlagBool(cmd, "server-side")
+type ServerSideApplyStrategy string
+
+const (
+	ServerSide ServerSideApplyStrategy = "true"
+
+	ClientSide ServerSideApplyStrategy = "false"
+
+	ServerSideAuto ServerSideApplyStrategy = "auto"
+)
+
+func GetServerSideApplyStrategy(cmd *cobra.Command) (ServerSideApplyStrategy, error) {
+	var serverSideFlag = GetFlagString(cmd, "server-side")
+	invalidOpt := fmt.Errorf(`invalid value %q for "--server-side" flag: value must be a boolean`, serverSideFlag)
+	if AutoServerSide.IsEnabled() {
+		invalidOpt = fmt.Errorf(`invalid value %q for "--server-side" flag: value must be a boolean or auto`, serverSideFlag)
+	}
+
```

Comment:
I would probably do some premapping here e.g.

```
switch strings.ToLower(serverSideFlag) {
  case "":
    // We do not want to change the default behaviour
    if flag.IsSet() {
      return ServerSideServer, nil
    } else {
      return ServerSideClient, nil
    }
  case "auto":
     if AutoServerSide.IsEnabled()  {
       return ServerSideAuto, nil
    }
}
```

(I don't know if we want to be case sensitive or not)

Reply:
I ended up going with the method used by the `dry-run` flag - setting the value for an unset flag in the add flag method, which cleans this up nicely. Then I double checked and the current behaviour is to fail on the flag being set to an empty string, so I maintained that.